### PR TITLE
patchwork: Init at 3.10.1

### PR DESCRIPTION
**Pull request by picnoir**
###### Motivation for this change

Add patchwork, a decentralized messaging and sharing app built on top of Secure Scuttlebutt (SSB).

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Hi,

As you can see, this software is distributed using a "new" (at least to me) archive type called [AppImage](https://appimage.org/). As its name states, it runs everywhere, it can install any software on any linux distribution, which sounds great.

Sadly, it does not run on nixos at all :(

I did not wanted to debug the "executable archive" by itself. Instead, I've looked for a way to extract the actual package from the archive in order to install it in a more "traditional" way.

I am extracting the actual package contained in various squashfs embedded in the AppImage using binwalk.

It's a bit hack-ish, but it has the certain advantage to work. I'm open to any improvement on the process.

I've setup a notification to watch any new release of this software: I plan to maintain this package.

Have a nice day!

### CODE REVIEW

**Comment by cryptix** [comment]
Hey! I'm very familiar with the scuttleverse and would take over the maintainance on this from ninja.


I'm also working on a way to patch ssb' electron apps that doesn't involve AppImages but I need some help with node2nix but that can happen on another discussion.

**Comment by NinjaTrappeur** [comment]
Added @cryptix as maintainer and squashed the commits.

**Review by samueldr — CHANGES_REQUESTED** [review]
I have left technical issues as comments in the files that will or may need resolving with the contribution.

**Comment by samueldr** [comment]
Additionally, while this is isn't the first AppImage that got contributed in nixpkgs, and definitely not the first binary build of a project, it is **always preferred** to get a source build going instead of going from a binary build.

Binary builds are generally used when

 * No sources are available (obviously).
 * The binaries have technical merits over a source build (partially open source software, of validated software where source builds differ.)
 * The build system is a mess hampering build from happening (e.g. some Java apps).

I haven't looked closely, but it [looks like the README has build instructions](https://github.com/ssbc/patchwork/blob/master/README.md).

One could also look at how other distributions (and unix-likes) package theirs:

 * https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=ssb-patchwork
 
(Uh, looks like only archlinux has a package. Let's make a second good example of a source-built one!)

I'm not 100% against using AppImages for nixpkgs, but this is probably a discussion better held elsewhere than in a contributions for a specific package. If another contributor feels it's good to merge this (once the misc. cleanups are applied) I won't hold it against them, but as this *looks* like it could be compiled from source, I strongly lobby for source compilation instead of using the binary build.

**Comment by samueldr** [comment]
I will additionally add that it may be of interest to synchronize efforts with regards to AppImage apps with the other contributor(s) contributing AppImage-based builds. A common framework to extract and make use of AppImages will probably be of value for nixpkgs even if all the packages using AppImage builds end up using source builds; AppImages distribute more than open source packages.

 See #44511.

**Comment by NinjaTrappeur** [comment]
Thank you very much for your review.

I privileged the appimage way because I personally had some pretty annoying experience of npm-based derivation randomly breaking because of some loose upper/lower dependencies bounds in the past.

@cryptix is planning to implement a source-based derivation in a near future. 

I don't think we need two duplicate patchwork derivations in the nixpkgs. I'm just gonna keep this one in a overlay for now and close this PR. Is this ok for you @samueldr ?

**Pull request closed without merging**
